pkg/agent/util/iptables/iptables.go

```diff
@@ -140,13 +140,17 @@ func (c *Client) EnsureRule(table string, chain string, ruleSpec []string) error
 			return fmt.Errorf("error checking if rule %v exists in table %s chain %s: %v", ruleSpec, table, chain, err)
 		}
 		if exist {
-			return nil
+			continue
```

Comment:
I don't understand the implementation of this function (before & after the fix). What if the rule is protocol-specific? Shouldn't this function be more like `InsertRule` below" take the protocol (v4 / v6) as a parameter and only consider the iptables matching this protocol?

Reply:
This `EnsureRule` method adds rule for all available iptables in this cluster (v4, v6 or both). While `InsertRule` adds rule according to the input. I think they are by design to be 2 different methods.

For `EnsureRule`, you can check code here:
https://github.com/antrea-io/antrea/blob/cac80840b1d9cd2f1b2cf887a4c570cf6864e0aa/pkg/agent/route/route_linux.go#L261
If it is an IPv6 only cluster, then there will only be one iptable for IPv6. The for loops for once. If it is a dual-stack one, then twice.

I think the bug here is that if it is dual-stack cluster, IPv4 rule exists and then it skip the check for IPv6. Maybe the author thnks that v4 and v6 rules are added or deleted together but without this change, traffic to hostnetwork Pod is dropped.

Reply:
I don't know if @tnqn has more background on this but this is confusing. I think the function comment should describe the behavior, and we may want to rename the function to something like `EnsureRuleInAllTables`.

Reply:
I was also confused by the current code when I recently made change to it. 
1. If there is an IP match in the rule, installing the rule for both IPv4 and IPv6 won't make sense.
2. The inconsistent case @lzhecheng mentioned, the issue may apply to `ChainExists` as well.

Reply:
Then, I think we should use a method like `InsertRule` to deal with IP match case, while `EnsureRule` (maybe renaming) for the rest.

Reply:
Still not convinced about the solution. Why don't we have single methods: `AddRule` / `AddChain`, which always take the protocol as a parameter, but possible protocol values are `"All"`, `"IPv4"` and `"IPv6"`?

Reply:
Used the Protocol struct in antrea's iptables.go as a parameter.